### PR TITLE
added option to use rtctstats from lib-jitsi-meet

### PR DESCRIPTION
**Pull request by nils-ohlmeier**
This PR adds a new option to use rtcstats from lib-jitsi-meet instead of the build in version in jitsi-meet.

### CODE REVIEW

**Comment by saghul** [comment]
Why don't we transition to using rtcstats in LJM in a single shot?

**Comment by saghul** [comment]
But wouldn't we roll back JM and LJM at the same time?

**Comment by nils-ohlmeier** [comment]
> But wouldn't we roll back JM and LJM at the same time?

Sorry I used the wrong word here. This allows us to turn the feature off with a config change and not have to do any software roll backs. I prefer config options over software rollbacks.
The downside is that once everything has been successfully transitioned over to the LJM version some code in JM should get removed.

File: react/features/rtcstats/RTCStats.ts [hunk 1]
```diff
@@ -103,29 +105,50 @@ class RTCStats {
...
+            const rtcOptions = {
+                meetingName: meetingFqn,
+                eventCallback: this.handleRtcstatsEvent.bind(this),
+                onCloseCallback: this.handleTraceWSClose.bind(this)
             };
 
-            const statsEntryCallback = { statsEntry: this.statsEntry.bind(this) };
+            JitsiMeetJS.setRtcStatsMeetingFqn(rtcOptions);
+
+            JitsiMeetJS.getRtcStatsTrace().then((traceF: any) => {
```
Comment (andrei-gavrilescu) [inline react/features/rtcstats/RTCStats.ts:121]: nit. await would work as well

File: react/features/rtcstats/middleware.ts
[the remark this answers is not in the document]
Comment (nils-ohlmeier) [reply react/features/rtcstats/middleware.ts:1]: Yes that is valid point.

Looking at middleware.js (mostly for later book keeping):

- LIB_WILL_INIT is basically replicated in the new LJM rtcstats code.
- CONFERENCE_JOINED and CONFERENCE_UNIQUE_ID_SET do call `connectAndSendIdentity()` which is problem you pointed out.
- TRACK_ADDED and TRACK_UPDATED send face marks which I believe is customization for Jitsi Meet, but also calls `sendVideoTypeData()` which I'm not sure if LJM users would need to do that as well.
- DOMINANT_SPEAKER_CHANGED, E2E_RTT_CHANGED, ADD_FACE_LANDMARKS are all customization for Jitsi Meet.
- CONFERENCE_TIMESTAMP_CHANGED I'm not sure if that needs to replicated by/for LJM users.
- CONFERENCE_WILL_LEAVE probably needs to be called by LJM users.

So I don't think everything in middleware.ts needs to replicated, but yes we probably need a second push taking care of the identity and at least the WILL_LEAVE pieces.

File: react/features/rtcstats/RTCStats.ts (hunk 1 above)
Comment (nils-ohlmeier) [reply react/features/rtcstats/RTCStats.ts:121]: Point taken. But TBH I'm not convinced that in this case, even though it would be less characters,  it would make the code more readable.

Please let me know if you want me to change it to await.

File: config.js [hunk 2]
```diff
@@ -992,6 +992,10 @@ var config = {
         // all SDPs with an empty string instead.
         // rtcstatsSendSdp: false,
 
+        // This determines if rtcstats from lib-jitsi-meet will be used instead of the one
+        // here in jitsi-meet directly
+        // rtcstatsUseLibJitsi: false,
```
Comment (saghul) [inline config.js:997]: Let's move this to `testing` since it's temporary...

File: react/features/rtcstats/RTCStats.ts [hunk 3]
```diff
@@ -103,29 +105,51 @@ class RTCStats {
...
-                pollInterval,
-                useLegacy,
-                sendSdp,
-                eventCallback: this.handleRtcstatsEvent.bind(this)
+        const { useLibRtcStats, meetingFqn } = options;
+
+        if (useLibRtcStats) {
+            logger.info('setting up libRtcStats');
+
+            JitsiMeetJS.rtcstats.events.on(
```
Comment (saghul) [inline react/features/rtcstats/RTCStats.ts:113]: THe `rtcstats` object itself should extend from EventEmitter.

File: config.js (hunk 2 above)
Comment (nils-ohlmeier) [reply config.js:997]: Done

Comment (nils-ohlmeier) [reply config.js]: Done.

File: react/features/rtcstats/RTCStats.ts (hunk 3 above)
Comment (nils-ohlmeier) [reply react/features/rtcstats/RTCStats.ts:113]: I think you are from an esthetic point of view right.

But the only way to accomplish this appears to be to have the new RtcStats class extend the EventEmitter class. Which then means RtcStats needs to be manually instantiated somewhere (I'm not sure where is the best point to do that). I failed yesterday to accomplish this, so that the Jitsi-meet code can actually hook up the `on` even listeners.

Long story short: I don't have the time left to accomplish this.

**Comment by saghul** [comment]
Dropping the approach, we'll migrate in a single shot.
